flow/aggregates.py

```diff
@@ -0,0 +1,481 @@
+        # "non-aggregate" operation.
+        if sort_by is None and select is None and not reverse_order:
+            self._is_aggregate = getattr(aggregator_function, '_num', 0) != 1
+        else:
+            self._is_aggregate = True
+
+        self._kwargs = kwargs
+        self._aggregator_function = aggregator_function
+        self._sort_by = sort_by
+        self._reverse_order = bool(reverse_order)
+        self._select = select
+
+    @classmethod
+    def groupsof(cls, num=1, sort_by=None, reverse_order=False, select=None):
+        """Decorator for operation function that aggregates signac jobs in groups of
+        a number provided by the user.
+
+        By default aggregate of a single job is created.
+
+        If the number of jobs present in the project can't be divided by the number
+        provided by the user then the last aggregate will have the maximum possible
+        jobs in it.
+        For instace, if 10 jobs are present in a project and they are aggregated in groups
+        of 3 then the length of aggregates would be 3, 3, 3, 1.
+
+        The below code-block provides an example on how jobs can be aggregated in
+        groups of 2.
+
+        .. code-block:: python
+
+            @aggregator.groupsof(num=2)
+            @FlowProject.operation
+            def foo(*jobs):
+                print(len(jobs))
+
+        :param num:
+            The default size of aggregates excluding the final aggregate.
+        :type num:
+            int
+        :param sort_by:
+            Before aggregating, sort the jobs by a given statepoint parameter.
+            The default behavior is no sorting.
+        :type sort_by:
+            str or NoneType
+        :param reverse_order:
+            States if the jobs are to be sorted in reverse order.
+            The default value is False.
+        :type reverse_order:
+            bool
+        :param select:
+            Condition for filtering individual jobs. This is passed as the
+            callable argument to `filter`.
+            The default behavior is no filtering.
+        :type select:
+            callable or NoneType
+        """
+        try:
+            if num != int(num):
+                raise ValueError('The num parameter should be an integer')
+            num = int(num)
+            if num <= 0:
+                raise ValueError('The num parameter should have a value greater than 0')
+        except TypeError:
+            raise TypeError('The num parameter should be an integer')
+
+        def aggregator_function(jobs):
+            args = [iter(jobs)] * num
+            return zip_longest(*args)
+
+        setattr(aggregator_function, '_num', num)
+
+        return cls(aggregator_function, sort_by, reverse_order, select, num=num)
+
+    @classmethod
+    def groupby(cls, key, default=None, sort_by=None, reverse_order=False, select=None):
+        """Decorator for operation function that aggregates signac jobs by grouping them
+        via a key.
```

Comment:
```suggestion
        """Aggregates jobs according to matching state point key values.
```

Reply:
the key can also be an arbitrary callable hence, according to me, users may get confused with this doc